services/consensusalgo/benchmarkconsensus/test/leader_test.go

```diff
@@ -6,11 +6,121 @@ import (
 	"testing"
 )
 
-func TestLeaderProposesBlocks(t *testing.T) {
+func newLeaderHarnessAndInit(t *testing.T, ctx context.Context) *harness {
+	h := newHarness(true)
+	h.expectNewBlockProposalNotRequested()
+	h.expectCommitSent(0, h.config.NodePublicKey())
+	h.createService(ctx)
+	h.verifyCommitSent(t)
+	return h
+}
+
+func TestLeaderInit(t *testing.T) {
+	test.WithContext(func(ctx context.Context) {
+		h := newLeaderHarnessAndInit(t, ctx)
+		h.verifyNewBlockProposalNotRequested(t)
+		h.verifyHandlerRegistrations(t)
+	})
+}
+
+func TestLeaderCommitsNextBlockAfterEnoughConfirmations(t *testing.T) {
+	test.WithContext(func(ctx context.Context) {
+		h := newLeaderHarnessAndInit(t, ctx)
+
+		h.expectNewBlockProposalRequestedAndSaved(1)
+		h.expectCommitSent(1, h.config.NodePublicKey())
+		h.receivedCommittedViaGossipFromSeveral(3, 0, true, true)
+		h.verifyNewBlockProposalRequestedAndSaved(t)
+		h.verifyCommitSent(t)
+
+		h.expectNewBlockProposalRequestedAndSaved(2)
+		h.expectCommitSent(2, h.config.NodePublicKey())
+		h.receivedCommittedViaGossipFromSeveral(3, 1, true, true)
+		h.verifyNewBlockProposalRequestedAndSaved(t)
+		h.verifyCommitSent(t)
+	})
+}
+
+func TestLeaderRetriesCommitOnError(t *testing.T) {
+	test.WithContext(func(ctx context.Context) {
+		h := newLeaderHarnessAndInit(t, ctx)
+
+		h.expectNewBlockProposalRequestedToFail()
+		h.expectCommitNotSent()
+		h.receivedCommittedViaGossipFromSeveral(3, 0, true, true)
+		h.verifyNewBlockProposalRequestedAndNotSaved(t)
+		h.verifyCommitNotSent(t)
+	})
+}
+
+func TestLeaderRetriesCommitAfterNotEnoughConfirmations(t *testing.T) {
+	test.WithContext(func(ctx context.Context) {
+		h := newLeaderHarnessAndInit(t, ctx)
+
+		h.expectNewBlockProposalRequestedAndSaved(1)
+		h.expectCommitSent(1, h.config.NodePublicKey())
+		h.receivedCommittedViaGossipFromSeveral(3, 0, true, true)
+		h.verifyNewBlockProposalRequestedAndSaved(t)
+		h.verifyCommitSent(t)
+
+		h.expectNewBlockProposalNotRequested()
+		h.expectCommitSent(1, h.config.NodePublicKey())
+		h.receivedCommittedViaGossipFromSeveral(2, 1, true, true)
+		h.verifyNewBlockProposalNotRequested(t)
+		h.verifyCommitSent(t)
+	})
+}
+
+func TestLeaderIgnoresBadSignatures(t *testing.T) {
+	test.WithContext(func(ctx context.Context) {
+		h := newLeaderHarnessAndInit(t, ctx)
+
+		h.expectNewBlockProposalNotRequested()
+		h.expectCommitSent(0, h.config.NodePublicKey())
+		h.receivedCommittedViaGossipFromSeveral(3, 0, false, true)
+		h.verifyNewBlockProposalNotRequested(t)
+		h.verifyCommitSent(t)
```

Comment:
why do we verify that commit was sent? shouldn't we ignore the bad signatures?

Reply:
or is this an assertion on the gossip messages? if so, why are you even asserting on that?